rclpy/src/rclpy/_rclpy.c

```diff
@@ -3369,6 +3370,299 @@ rclpy_clock_set_ros_time_override(PyObject * Py_UNUSED(self), PyObject * args)
+        return NULL;
+      }
+      PyList_SET_ITEM(value, i, member_value);
+    }
+  } else if (variant->double_array_value) {
+    type_enum_value = 8;
+    value = PyList_New(variant->double_array_value->size);
+    if (NULL == value) {
+      return NULL;
+    }
+    for (size_t i = 0; i < variant->double_array_value->size; i++) {
+      member_value = PyFloat_FromDouble(variant->double_array_value->values[i]);
+      if (NULL == member_value) {
+        Py_DECREF(value);
+        return NULL;
+      }
+      PyList_SET_ITEM(value, i, member_value);
+    }
+  } else if (variant->string_array_value) {
+    type_enum_value = 9;
+    value = PyList_New(variant->string_array_value->size);
+    if (NULL == value) {
+      return NULL;
+    }
+    for (size_t i = 0; i < variant->string_array_value->size; i++) {
+      member_value = PyUnicode_FromString(variant->string_array_value->data[i]);
+      if (NULL == member_value) {
+        Py_DECREF(value);
+        return NULL;
+      }
+      PyList_SET_ITEM(value, i, member_value);
+    }
+  }
+
+  PyObject * args = Py_BuildValue("(i)", type_enum_value);
+  if (NULL == args) {
+    return NULL;
+  }
+  PyObject * type = PyObject_CallObject(parameter_type_cls, args);
+  Py_DECREF(args);
+  args = Py_BuildValue("sOO", name, type, value);
+  if (NULL == args) {
+    return NULL;
+  }
+  PyObject * param = PyObject_CallObject(parameter_cls, args);
+  Py_DECREF(args);
+  Py_DECREF(type);
+  return param;
+}
+
+/// Get a list of parameters for the current node from rcl_yaml_param_parser
+/**
+ * On failure, an exception is raised and NULL is returned if:
+ *
+ * Raises ValueError if the argument is not a node handle.
+ * Raises RuntimeError if the parameters file fails to parse
+ *
+ * \param[in] parameter_cls The rclpy.parameter.Parameter class object.
+ * \param[in] node_handle Capsule pointing to the node handle
+ * \return NULL on failure
+ *         A list of rclpy.parameter.Parameter on success (may be empty).
+ *
+ */
+static PyObject *
+rclpy_get_node_parameters(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  PyObject * parameter_cls;
+  PyObject * node_capsule;
+  if (!PyArg_ParseTuple(args, "OO", &parameter_cls, &node_capsule)) {
+    return NULL;
+  }
+
+  rcl_node_t * node = (rcl_node_t *)PyCapsule_GetPointer(node_capsule, "rcl_node_t");
+  if (!node) {
+    return NULL;
+  }
+
+  const rcl_node_options_t * node_options = rcl_node_get_options(node);
+  const rcl_allocator_t allocator = node_options->allocator;
+  rcl_params_t * params = rcl_yaml_node_struct_init(allocator);
+  if (NULL == params) {
+    PyErr_Format(PyExc_RuntimeError, "Failed to allocate initial parameters");
+    return NULL;
+  }
+
+  if (node_options->use_global_arguments) {
+    if (!_parse_param_files(rcl_get_global_arguments(), allocator, params)) {
+      return NULL;
+    }
+  }
+
+  if (!_parse_param_files(&(node_options->arguments), allocator, params)) {
+    return NULL;
+  }
+
+  int node_index = -1;
```

Comment:
MSBuild (rightly) complains that `int`might lose information when assigned from a `size_t` below. But changing to a `size_t` makes the sigil value `-1` invalid. Do we have a pattern we use for this sort of situation like storing a second node_index_found value and setting and checking it along with the index?